File: typescript/chapters/ch10/src/core/profiles.ts
export type Capability =
  | "loop"
  | "powershell"
  | "tool_registry"
  | "files"
  | "policy"
  | "hooks"
  | "todo"
  // subagent 表示可将自包含工作交给隔离 AgentRunner 执行。
  | "subagent"
  // skills 表示可以发现 workspace 技能摘要并按名称加载受控说明。
  | "skills"
  // artifacts 与 compaction 共同表示大结果落盘和受预算的上下文缩减。
  | "artifacts"
  | "compaction"
  // memory 表示回合可检索、提取并持久化工作区级记忆。
  | "memory"
  // dynamic_prompt 表示系统提示由运行时状态渲染，而非固定字符串。
  | "dynamic_prompt";

export interface ChapterProfile {
  readonly chapter: number;
  readonly capabilities: ReadonlySet<Capability>;
}

// 用不可导出底层 Set 的包装实现 ReadonlySet，防止 profile 能力被外部修改。
class CapabilitySet implements ReadonlySet<Capability> {
  readonly #values: Set<Capability>;

  constructor(values: readonly Capability[]) {
    this.#values = new Set(values);
  }

  get size(): number {
    return this.#values.size;
  }

  has(value: Capability): boolean {
    return this.#values.has(value);
  }

  entries(): SetIterator<[Capability, Capability]> {
    return this.#values.entries();
  }

  keys(): SetIterator<Capability> {
    return this.#values.keys();
  }

  values(): SetIterator<Capability> {
    return this.#values.values();
  }

  forEach(
    callbackfn: (value: Capability, value2: Capability, set: ReadonlySet<Capability>) => void,
    thisArg?: unknown,
  ): void {
    this.#values.forEach((value) => {
      callbackfn.call(thisArg, value, value, this);
    });
  }

  [Symbol.iterator](): SetIterator<Capability> {
    return this.values();
  }
}

export const P01: ChapterProfile = Object.freeze({
  chapter: 1,
  capabilities: new CapabilitySet(["loop", "powershell"]),
});

export const P02: ChapterProfile = Object.freeze({
  chapter: 2,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files"]),
});

export const P03: ChapterProfile = Object.freeze({
  chapter: 3,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files", "policy"]),
});

// P04 在 P03 基础上加入 hooks，允许受限 Hook 生命周期扩展。
export const P04: ChapterProfile = Object.freeze({
  chapter: 4,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
  ]),
});

// P05 在 P04 基础上加入 todo，用于会话级计划快照与陈旧提醒。
export const P05: ChapterProfile = Object.freeze({
  chapter: 5,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
  ]),
});

// P06 在 P05 基础上加入 subagent，用于隔离委派并复用父级审批边界。
export const P06: ChapterProfile = Object.freeze({
  chapter: 6,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
  ]),
});

// P07 在 P06 基础上增加 skills 能力，让组合根按需扫描并注册知识加载工具。
export const P07: ChapterProfile = Object.freeze({
  chapter: 7,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
  ]),
});

// P08 在 P07 基础上增加 artifacts 与 compaction，让组合根接入请求级压缩和结果落盘。
export const P08: ChapterProfile = Object.freeze({
  chapter: 8,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
  ]),
});

// P09 在 P08 能力上追加 memory，使同一 Loop 同时拥有请求级压缩与跨会话文件记忆。
export const P09: ChapterProfile = Object.freeze({
  chapter: 9,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
  ]),
});

// P10 在 P09 能力上追加 dynamic_prompt，让组合根切换到运行时渲染系统提示。
export const P10: ChapterProfile = Object.freeze({
  // P10 将系统提示改为运行时渲染，以便工具、Skill 和记忆状态即时反映。
  chapter: 10,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
    "dynamic_prompt",
  ]),
});

// 仅返回模块内冻结的单例，供组合根用引用相等性校验 profile 来源。
export function profileForChapter(chapter: number): ChapterProfile {
  if (chapter === 1) {
    return P01;
  }
  if (chapter === 2) {
    return P02;
  }
  if (chapter === 3) {
    return P03;
  }
  if (chapter === 4) {
    return P04;
  }
  if (chapter === 5) {
    return P05;
  }
  if (chapter === 6) {
    return P06;
  }
  if (chapter === 7) {
    return P07;
  }
  if (chapter === 8) {
    return P08;
  }
  if (chapter === 9) {
    return P09;
  }
  if (chapter === 10) {
    return P10;
  }
  if (!Number.isInteger(chapter) || chapter < 1 || chapter > 20) {
    throw new Error("chapter must be an integer from 1 to 20");
  }
  throw new Error(`Chapter ${chapter} has not been migrated to TypeScript yet`);
}

File: python/ch09_agent/agent_ch09/core/profiles.py
"""固定章节能力快照。

Java 对照：`ChapterProfile` 类似不可变配置 record，P01-P09 类似预定义单例常量。
组合根使用对象身份判断，调用方不能临时拼一个同字段对象来冒充正式章节。

这是什么：章节能力的配置定义，每章渐进式增加功能
Java 类比：类似 Spring Profile 或枚举配置类
为什么需要：确保每章只启用对应的功能，防止能力越级使用
"""

from dataclasses import dataclass
from typing import Literal

# 能力类型：定义所有可用的 Agent 功能模块
Capability = Literal[
    "loop",         # 核心循环
    "powershell",   # PowerShell 命令执行
    "tool_registry",# 工具注册表
    "files",        # 文件系统操作
    "policy",       # 权限策略
    "hooks",        # Hook 生命周期
    "todo",         # TODO 跟踪
    "subagent",     # 子 Agent
    "skills",       # Skill 按需加载
    "artifacts",    # Artifact 支持
    "compaction",   # 上下文压缩
    "memory",       # 长期记忆
]


@dataclass(frozen=True, slots=True)
class ChapterProfile:
    """章节号及其允许装配的能力白名单。

    这是什么：章节配置的不可变数据对象
    Java 类比：类似 record ChapterProfile(int chapter, Set<Capability> capabilities)
    为什么需要：定义每章的能力边界，确保教学进度和功能匹配

    参数：
        chapter: 章节号（1-9）
        capabilities: 该章允许使用的能力集合（不可变）
    """

    chapter: int  # 章节号
    capabilities: frozenset[Capability]  # 能力白名单


# ==================== 章节配置单例 ====================
# 每个章节是一个预定义的不可变对象，通过对象身份（is）判断

P01 = ChapterProfile(1, frozenset({"loop", "powershell"}))
"""第一章：基础循环 + PowerShell 命令执行"""

P02 = ChapterProfile(2, frozenset({"loop", "powershell", "tool_registry", "files"}))
"""第二章：增加工具注册表和文件系统操作"""

P03 = ChapterProfile(3, frozenset({"loop", "powershell", "tool_registry", "files", "policy"}))
"""第三章：增加权限策略"""

P04 = ChapterProfile(
    4, frozenset({"loop", "powershell", "tool_registry", "files", "policy", "hooks"})
)
"""第四章：增加 Hook 生命周期"""

P05 = ChapterProfile(
    5, frozenset({"loop", "powershell", "tool_registry", "files", "policy", "hooks", "todo"})
)
"""第五章：增加 TODO 跟踪"""

P06 = ChapterProfile(
    6,
    frozenset(
        {"loop", "powershell", "tool_registry", "files", "policy", "hooks", "todo", "subagent"}
    ),
)
"""第六章：增加子 Agent"""

P07 = ChapterProfile(
    7,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
        }
    ),
)
"""第七章：增加 Skill 按需加载"""

P08 = ChapterProfile(
    8,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
            "artifacts",
            "compaction",
        }
    ),
)
"""第八章：增加 Artifact 支持和上下文压缩"""

P09 = ChapterProfile(
    9,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
            "artifacts",
            "compaction",
            "memory",
        }
    ),
)
"""第九章：增加长期记忆"""


def profile_for_chapter(chapter: int) -> ChapterProfile:
    """根据章节号返回对应的配置对象。

    这是什么：章节配置的查找方法
    Java 类比：类似 ChapterProfile getProfile(int chapter)
    为什么需要：提供统一的配置获取接口，确保只返回预定义的单例

    参数：
        chapter: 章节号（1-9）

    返回：
        ChapterProfile: 对应章节的配置单例

    异常：
        ValueError: 章节号无效或尚未迁移
    """
    # 章节号到配置对象的映射表
    profiles = {1: P01, 2: P02, 3: P03, 4: P04, 5: P05, 6: P06, 7: P07, 8: P08, 9: P09}

    # 校验章节号类型和范围
    if isinstance(chapter, bool) or not isinstance(chapter, int) or not 1 <= chapter <= 20:
        raise ValueError("chapter 必须是 1 到 20 的整数")

    # 检查章节是否已迁移
    if chapter not in profiles:
        raise ValueError(f"第 {chapter} 章尚未迁移为 Python")

    return profiles[chapter]
